File: 프로그래머스/lv3/118668. 코딩 테스트 공부/코딩 테스트 공부.java
import java.util.*;

class Solution {
    final int INF = 987654321;
    
    int maxAlp = 0;
    int maxCop = 0;
    
    int[][] dp;
    
    int dfs(int alp, int cop, List<int[]> list) {
        if(alp == maxAlp && cop == maxCop) {
            return 0;
        }
        
        if(dp[alp][cop] != INF) {
            return dp[alp][cop];
        }
        
        dp[alp][cop] = INF + 1;
                
        for(int i=0;i<list.size();i++) {
            int[] problem = list.get(i);
            
            if(alp >= problem[0] && cop >= problem[1]) {
                int nextAlp = Math.min(alp + problem[2], maxAlp);
                int nextCop = Math.min(cop + problem[3], maxCop);
                
                dp[alp][cop] = Math.min(dp[alp][cop],
                                    dfs(nextAlp, nextCop, list) + problem[4]);
            }
        }
        
        return dp[alp][cop];
    }
    
    public int solution(int alp, int cop, int[][] problems) {
        List<int[]> list = new ArrayList<>();
        
        for(int i=0;i<problems.length;i++) {
            maxAlp = Math.max(maxAlp, problems[i][0]);
            maxCop = Math.max(maxCop, problems[i][1]);
            
            list.add(problems[i]);
        }
        
        maxAlp = Math.max(alp, maxAlp);
        maxCop = Math.max(cop, maxCop);
        
        list.add(new int[]{0, 0, 1, 0, 1});
        list.add(new int[]{0, 0, 0, 1, 1});
        
        dp = new int[maxAlp + 1][maxCop + 1];
        
        for(int i=alp;i<=maxAlp;i++) {
            for(int j=cop;j<=maxCop;j++) {
                dp[i][j] = INF;
            }
        }
        
        return dfs(alp, cop, list);
    }
}

File: 프로그래머스/Lv.3/118668. 코딩 테스트 공부/코딩 테스트 공부.py
INF = 1000
maxAlp = 0
maxCop = 0

def dfs(alp, cop, problems, dp):
    if alp >= maxAlp and cop >= maxCop:
        return 0
    
    if dp[alp][cop] != INF:
        return dp[alp][cop]
    
    if alp < maxAlp:
        dp[alp][cop] = min(dp[alp][cop], dfs(alp + 1, cop, problems, dp) + 1)
    
    if cop < maxCop:
        dp[alp][cop] = min(dp[alp][cop], dfs(alp, cop + 1, problems, dp) + 1)
    
    for problem in problems:
        if alp >= problem[0] and cop >= problem[1]:
            dp[alp][cop] = min(dp[alp][cop], dfs(min(maxAlp, alp + problem[2]), min(maxCop, cop + problem[3]), problems, dp) + problem[4])
    
    return dp[alp][cop]

def solution(alp, cop, problems):
    global maxAlp, maxCop
    
    for problem in problems:
        maxAlp = max(maxAlp, problem[0])
        maxCop = max(maxCop, problem[1])
    
    dp = [[INF for _ in range(maxCop + 1)] for _ in range(maxAlp + 1)]
    
    return dfs(min(maxAlp, alp), min(maxCop, cop), problems, dp)